mod_form.php

```diff
@@ -292,4 +300,24 @@ public function add_completion_rules() {
     public function completion_rule_enabled($data) {
         return true;
     }
+    public function validation($data, $files) {
+        $errors = parent::validation($data, $files);
+
+        if (!empty($data['emailperiod']) && $data['emailperiod'] <= 60 && $data['emailfrequency'] > 5) {
+            // Email is sent based on minutes selector - make sure no more than 5 emails can be sent.
+            $errors['emailfrequency'] = get_string('frequencytoohigh', 'reengagement', 5);
```

Comment:
I like your comprimse here re spam restrictions - avoiding the whole opt-out complexity & leaving it up to the course creator to sort out.

But I'd argue for a more restrictive threshold.  Allowing someone to spam 5 times over 25 minutes, or 12 times over a day seems a bit much.